### PR TITLE
Fix Merlin per-module pp flags

### PR DESCRIPTION
**Pull request by voodoos**
Recent changes made merlin configuration files distinct for each module in a lib of exe. However we were not taking advantage on that for preprocessing, this PR correct that.

This should fix issues: #2596 #1212 and #3409

@rgrinberg this PR is based on #4091. I shared the "processing" work for every module using the same pp specification. However the shared `Per_item` data-structure does not play well with `Build.With_target.t` and is not easily translatable to a `(key list, 'a) list)` while preserving sharing. I decided to duplicate these once-processed pp flags for each module, does that seem reasonable to you ?

### CODE REVIEW

**Comment by voodoos** [comment]
The more I think about it the more I feel the duplication should not happen.

We just need some kind of `Build.With_targets.all` for `Per_item.t` and it would make for a much nicer implementation.

**Comment by voodoos** [comment]
The only option to keep sharing I see would be to add a function to `Per_element` because we need access to the array used in the implementation. The function could be something like that:

```
let map_with_targets { map; values } ~f =
  let open Build.With_targets.O in
  let l = Array.to_list values in
  let+ new_values = List.map l ~f |> Build.With_targets.all in
  {
    map;
    values = Array.of_list new_values
  }
```

**Comment by rgrinberg** [comment]
>  I shared the "processing" work for every module using the same pp specification. However the shared Per_item data-structure does not play well with Build.With_target.t and is not easily translatable to a (key list, 'a) list)

That is a problem for all of our containers. They require this projection and projection to lists to go from `'a Build.t container` to `'a container Build.t`.

>  I decided to duplicate these once-processed pp flags for each module, does that seem reasonable to you ?

Sadly, I don't think this is good enough.

> The only option to keep sharing I see would be to add a function to Per_element because we need access to the array used in the implementation. The function could be something like that:

That sounds much better to me.

By the way, we should maintain the `Per_item` representation when marshalling. We don't want to write bloated marshalled files unnecessarily, the reader could do the  lookup inside the `Per_item` for the module of interest just as easily

**Comment by voodoos** [comment]
@rgrinberg : should be better like this. Sharing is preserved until the marshaled config is read and pp'ed by the `merlin-server`.

However the debug printing is duplicating again because the data-structure used in `Per_item` does not allow for the easy recovery of all members of a same equivalence class. But this is only for the debug print anyway.

**Comment by Khady** [comment]
> A fresh profile with this PR would be helpful.

- f51644afc20b96eeff103fd2dd9b0d9168ffa978 [trace.zip](https://github.com/ocaml/dune/files/5801510/trace.zip)
- e7a0e930391ea818e368a422337423b2e7f65d48 [trace.zip](https://github.com/ocaml/dune/files/5801580/trace.zip)

File: src/dune_rules/merlin.ml [hunk 1]
```diff
@@ -61,38 +60,43 @@ module Processed = struct
...
       (List.concat [ exclude_query_dir; obj_dirs; src_dirs; flags; suffixes ])
 
-  let rec get config ~filename =
-    let file = Filename.remove_extension filename in
-    match Option.map (String.Map.find config filename) ~f:to_sexp with
-    | Some result -> Some result
-    | None when String.equal file filename -> None
-    | None -> get config ~filename:file
+  let get { modules; pp_config; config } ~filename =
+    let fname = Filename.remove_extension filename in
```
Comment (rgrinberg) [inline src/dune_rules/merlin.ml:75]: I think this `fname` should probably be lower cased as well. Or, you could  just take compare it to one of the `Module.source` of the module.

File: src/dune_rules/merlin.ml [hunk 2]
```diff
@@ -61,38 +60,43 @@ module Processed = struct
...
+    match
+      List.find_opt modules ~f:(fun name ->
+          let fname' = Module_name.to_string name |> String.lowercase in
+          String.equal fname fname')
+    with
+    | Some name ->
+      let pp = Module_name.Per_item.get pp_config name in
+      Some (to_sexp ~pp config)
+    | None -> None
```
Comment (rgrinberg) [inline src/dune_rules/merlin.ml:84]: This can be replaced with `|> Option.map ~f:(fun name -> ..)`

File: src/dune_rules/merlin.ml [hunk 3]
```diff
@@ -3,29 +3,28 @@ open! Stdune
...
-  let to_dyn = Tuple.T2.to_dyn String.to_dyn String.to_dyn
-end)
-
 module Processed = struct
   (* The actual content of the merlin file as built by the [Unprocessed.process]
      function from the unprocessed info gathered through [gen_rules]. The first
      three fields map directly to Merlin's B, S and FLG directives and the last
      one represents a list of preprocessors described by a preprocessing flag
      and its arguments. *)
+  type pp = (string * string) option
```
Comment (rgrinberg) [inline src/dune_rules/merlin.ml:12]: Could you make this a record? It's not obvious what the tuple fields are.

**Review by rgrinberg — APPROVED** [review]
Looks good apart from some minor things.

**Comment by rgrinberg** [comment]
Thanks @Khady . Could you also give a `perf` report of the last commit here.

I can confirm that all merlin related slowness is now absent from the trace. Apart from the things I mentioned earlier, I don't have any ideas on what else we can optimize. Perhaps @aalekseyev will find something though.

**Comment by aalekseyev** [comment]
In the profile I couldn't see any obvious single thing to work on.
A lot of it seems to be overhead of Fiber and Memo, though. For example `add_dep_from_caller` seems to account for ~15% of all allocations, which is not what I would expect.

**Comment by rgrinberg** [comment]
@voodoos Ready to merge!

Actually this PR also fixes a bug along the way: directories with mix preprocessors specifications are now handled correctly. You should mention that in the CHANGES along with the appropriate bug numbers for it.

**Comment by Khady** [comment]
> Could you also give a perf report of the last commit here.

Here it is [perf.zip](https://github.com/ocaml/dune/files/5813618/perf.zip)

**Pull request merged**